R/vars-select.R

```diff
@@ -229,20 +243,36 @@ inds_combine <- function(vars, inds) {
 }
 
 inds_unique <- function(x) {
-  # Remove duplicates
-  out <- vctrs::vec_unique(x)
+  # Remove duplicates but keep last name of duplicates
+  split <- vctrs::vec_split(x, x)
+  out <- split$key
 
   # Keep last name of duplicates
   if (length(out) < length(x)) {
-    reversed <- rev(x)
-    rev_unique_locs <- vctrs::vec_unique_loc(reversed)
-    unique_nms <- rev(names2(reversed)[rev_unique_locs])
-    names(out) <- unique_nms
+    names(out) <- map_chr(split$val, ind_last_name)
   }
 
   out
 }
 
+ind_last_name <- function(x) {
+  names <- names(x)
+  last <- last(names)
+
+  if (!all(vctrs::vec_duplicate_detect(names) | names == "")) {
+    dups <- encodeString(names, quote = "\"")
+    dest <- last(dups)
+    dups <- glue::glue_collapse(dups, ", ", last = " and ")
+    msg <- glue::glue(
+      "Can't rename the same variable to different names ({dups}).
```

Comment:
Can we get the name of the variable being renamed here? Then the message could be:

```
Must rename a variable to a single choice: {var} is being renamed to {dups}. 
Using `{dest} = {var}`
```

Reply:
I like it. Maybe this would be even clearer this way?

```
Can't rename variables to multiple choices: `a` is being renamed to `d` and `e`.
Renaming to `e`. 
```

Reply:
The style guide suggests using must when we know what the correct result is, but I agree that my wording still seems clumsy. Maybe @batpigandme will have other ideas.

Reply:
I like the call out of `Can't rename variables to multiple choices` (it's clear what the action was that is forbidden), and/but `Must rename a variable to a single choice` is the corrective. 

Is the behaviour that it automatically renames to the last name given? If not, you could use the `Must rename variable to a single choice using {dest} = {var}`.

Reply:
I don't like "Must" here because I think it implies there should be an error rather than a warning. Here we "can't" rename to multiple choices, but we have taken action to continue nonetheless.

Reply:
Also how about

> Renaming `a` to `e`.

instead of

> Using `e = a`.

The latter will make users pause as they decipher the new = old syntax.

Reply:
Ah, your points are noted and make sense to me.

Reply:
WDYT of my last two comments @hadley?

Reply:
I no longer think that "must" is a problem for warnings. The message is now:

```
Must rename variables to a single choice: `a` is being renamed to `d` and `e`.
Renaming `a` to `e` instead.
```

However, I still find this message structure confusing. When we know both what can't happen and what must be done, I think I'd prefer:

```
Can't rename `a` to `d` and `e` because variables must be renamed to a single choice.
Renaming `a` to `e` instead.
```

I find it clearer because we start with the action that the user attempted rather than with what they should do instead. I think it makes it easier for the user to connect the warning to their code.